crates/dht/src/time.rs

```diff
@@ -0,0 +1,1051 @@
+        Self {
+            origin_timestamp,
+            factor,
+            full_slices: 0,
+            partial_slices: Vec::new(),
+            // This only changes based on the factor, which can't change at runtime,
+            // so compute it on construction.
+            // Note that `- 1` is because the first partial slice is half the size of a full slice,
+            // so we actually want to calculate to `factor - 1`.
+            min_recent_time: residual_duration_for_factor(factor - 1),
+            // Immediately requires an update, any time in the past will do
+            next_update_at: Timestamp::from_micros(0),
+        }
+    }
+
+    /// The timestamp at which the last full slice ends.
+    ///
+    /// If there are no full slices, then this is the origin timestamp.
+    fn full_slice_end_timestamp(&self) -> Timestamp {
+        let full_slices_duration = Duration::from_secs(
+            self.full_slices * (1u64 << self.factor) * UNIT_TIME.as_secs(),
+        );
+        self.origin_timestamp + full_slices_duration
+    }
+
+    /// Layout the partial slices for the given time range.
+    ///
+    /// Tries to allocate as many large slices as possible to fill the space.
+    fn layout_partials(
+        &self,
+        current_time: Timestamp,
+        mut start_at: Timestamp,
+    ) -> K2Result<Vec<(Timestamp, u8)>> {
+        let mut recent_time = (current_time - start_at).map_err(|_| {
+            K2Error::other("Failed to calculate recent time for partials, either the clock is is wrong or this is a bug")
+        })?;
+
+        let mut partials = Vec::new();
+
+        // Now we want to partition the remaining time into smaller slices
+        for i in (0..self.factor).rev() {
+            // Starting from the largest slice size, if there's space for two of that slice size
+            // then add two slices of that size, otherwise add one slice of that size
+            let slice_size =
+                Duration::from_secs((1u64 << i) * UNIT_TIME.as_secs());
+            let slice_count =
+                if recent_time > residual_duration_for_factor(i) + slice_size {
+                    2
+                } else if recent_time > slice_size {
+                    1
+                } else {
+                    continue;
+                };
+
+            for _ in 0..slice_count {
+                partials.push((start_at, i));
+                start_at += slice_size;
+                recent_time -= slice_size;
+            }
+        }
+
+        Ok(partials)
+    }
+}
+
+/// Computes what duration is required for a series of time slices.
+///
+/// The duration is computed as the sum of the powers of two from 0 to the factor, multiplied by
+/// the [UNIT_TIME].
+///
+/// Panics if the factor is greater than 53.
+/// Note that the maximum factor changes if the [UNIT_TIME] is changed.
+fn residual_duration_for_factor(factor: u8) -> Duration {
+    if factor >= 54 {
+        panic!("Factor must be less than 54");
```

Comment:
Ah, that's why. Maybe an explanation in the docs.

Reply:
Agreed, added a note about selecting a factor